Add a new error code for incorrectly closed connections

JDepooter opened the pull request:
I recently came across a situation in which curl would report an error when built with schannel or gnutls, but not when built with openssl. The server in question was closing the TLS connection without sending a TLS close notify alert.

The [schannel code](https://github.com/curl/curl/blob/master/lib/vtls/schannel.c#L1269) states that this should be an error, because it is a security issue related to SSL truncation attacks. At the same time, applications emphasizing a smoother user experience (ie browsers) ignore this condition. This is difficult to do with libcurl, as not all backends return an error, and those that do return a very generic error (CURLE_RECV_ERROR). This makes it hard to determine which errors are actual transfer errors, and which are less critical, such as the one described above.

I've added a new error code for this situation, which will let libcurl users decide how it should be handled in their application. I've implemented support for this condition in the schannel, openssl and gnutls backends.

I would also be open to renaming the new error code to something slightly more generic. This would allow the same error code to be reused in other cases in which the data transfer completed successfully, but an error occurred while closing the connection.

mback2k commented on lib/strerror.c:
```diff
@@ -308,6 +308,9 @@ curl_easy_strerror(CURLcode error)
     return "Stream error in the HTTP/2 framing layer";
 
+  case CURLE_MISSING_TLS_CLOSE_NOTIFY:
+     return "Server closed the TLS connection without a close notification";
```
Looks like there is a space indent too much in this line, please fix the indentation.

JDepooter replied:
Updated with the extra space removed.

mback2k commented:
The small indentation mistake aside, this pull-request looks good to me. :-)

mback2k commented:
It seems like test 1119 fails because of the new error code. Please take a look:
```
test 1119...[Verify that symbols-in-versions and headers are in sync]
perl  returned 2, when expecting 0
 exit FAILED
== Contents of files in the log/ dir after test 1119
=== Start of file stdout1119
 CURLE_MISSING_TLS_CLOSE_NOTIFY
=== End of file stdout1119
```

nickzman commented:
You can also implement the new error code in the Secure Transport back-end. In lib/vtls/darwinssl.c, in darwinssl_recv(), if SSLRead() returns errSSLClosedNoNotify, then that means that the server abruptly hung up without the notify alert. We look out for this at line 2482, but you can change that to return the new error.

JDepooter commented:
@mback2k I've updated the commits with your requested changes. Test 1119 is now passing.

@nickzman I've incorporated your suggested change as well. Can you please take a look, as I don't have a mac to test this on.

mback2k commented:
Looks good to me. @bagder Any final words? :-)

bagder commented:
My only thought is if we should make the naming more  generic so that it could be used for other forms of "protocol errors after all the data has arrived". Ideas?

mback2k commented:
Or maybe something like "unexpected connection close", because that is not TLS specific, but describes the situation pretty well?

bagder commented:
The close might not be unexpected but it is done "uncleanly" right?

"Incorrect connection termination"  perhaps?

mback2k commented:
Sounds good. @JDepooter could you please update your pull request accordingly?

jay commented:
Though the changes in schannel passed by without any complaint I'm not confident we'll see the same with OpenSSL. A ton of people are using it as the backend, and we can assume there are some servers that will close unclean.

bagder commented:
Right, I'm also a bit scared to find out how often this happens and thus how many people will now suddenly start to get errors when they previously had things returning OK... Can we do anything to smoothen that possible bump?

jay commented:
Does Mozilla have any telemetry on how often this happens?

bagder commented:
I don't think so. Not that I can spot on https://telemetry.mozilla.org, but I'm also not an expert on NSS and its integration in Firefox.

JDepooter commented:
@mback2k @bagder 
I have updated this pull request to use a more generic name: CURLE_INCORRECT_TERMINATION.

I have reverted the OpenSSL change for now, using a separate commit so that it can be easily retrieved in the future, if needed. My use case is actually to ignore the error (hence adding a new error code to distinguish from other errors), so I am fine with leaving OpenSSL as is for the moment.

bagder commented on docs/libcurl/symbols-in-versions:
```diff
@@ -88,6 +88,7 @@ CURLE_LDAP_SEARCH_FAILED        7.1
 CURLE_LIBRARY_NOT_FOUND         7.1           7.17.0
 CURLE_LOGIN_DENIED              7.13.1
 CURLE_MALFORMAT_USER            7.1           7.17.0
+CURLE_INCORRECT_TERMINATION     7.50.2
```
This should be `7.51.0` now!

bagder commented on docs/libcurl/libcurl-errors.3:
```diff
@@ -253,6 +253,9 @@ Failed to match the pinned key specified with \fICURLOPT_PINNEDPUBLICKEY(3)\fP.
 Stream error in the HTTP/2 framing layer.
+.IP "CURLE_INCORRECT_TERMINATION (93)"
+The server abruptly and unexpectedly closed the connection after a successful
+transfer. (Added in 7.50.2)
```
Added in `7.51.0`

jay commented:
I still think we should have some metrics on this before doing it. Apple has how many devices running curl/libcurl? What if even a small fraction of those connections start to fail. Maybe we should make it an option like CURLOPT_FAIL_PREMATURE_CLOSE. or shorter whatever, and default it to off.

Another thing is the text says server unexpectedly closed after successful transfer but something like GNUTLS_E_PREMATURE_TERMINATION afaics is not indicative of a successful transfer; we base that on content-length among other things for different protocols.

Is close_notify is strictly required? I read the [Closure Alerts section](https://tools.ietf.org/html/rfc5246#section-7.2.1) but if we know the length (or that we're waiting for a line ending) won't we know of truncation?

@bagder do you think Mozilla would consider collecting some anonymous metrics on this? how often no close notify is received and also is the content length known, http version etc

bagder commented:
Maybe it would make more sense if we instead go the other way: make sure that no TLS backend reports error if the entire contents has been received fine? After all, who exactly is it that will benefit from getting this new error where it previously returned OK?

> do you think Mozilla would consider collecting some anonymous metrics on this

I think they could be persuaded if A) there's an upside for Mozilla to get this information and B) someone writes up a patch for it unless (A) provides a very compelling motivation for Mozilla to hand over the job to paid staff. I'm not sure it is worth the rather hefty investment in time and effort for us.

JDepooter commented:
Very well, I will submit a new pull request which makes this a non-error situation rather than a unique error.
